File: ui/src/types/schemaRegistry.ts
export type ProposalType = "add" | "edit" | "delete";
export type ProposalStatus =
  | "pending"
  | "approved"
  | "rejected"
  | "apply_failed";
export type CriticalFor = "always" | "create_only";

export interface DatasetGroup {
  id: string;
  key: string;
  name: string;
  description: string | null;
}

export interface CreateDatasetGroupRequest {
  key: string;
  name: string;
  description?: string | null;
}

export interface SchemaDataset {
  id: string;
  key: string;
  group_id: string | null;
  column_count: number;
  pending_proposal_count: number;
}

export interface RegistryColumn {
  name: string;
  data_type: string;
  is_nullable: boolean | null;
  is_important: boolean | null;
  is_system_generated: boolean | null;
  description: string | null;
  primary_key: boolean | null;
  partition_order: number | null;
  license: string | null;
  units: string | null;
  hint: string | null;
  is_mandatory: boolean | null;
  is_unique: boolean | null;
  domain_values: string | null;
  range_min: number | null;
  range_max: number | null;
  dynamic_max: string | null;
  precision_min: number | null;
  critical_for: CriticalFor | null;
}

export interface CreateProposalRequest {
  column_name: string;
  proposal_type: ProposalType;
  after_state?: Partial<RegistryColumn> | null;
}

export interface Proposal {
  id: string;
  dataset_id: string;
  column_name: string;
  proposal_type: ProposalType;
  status: ProposalStatus;
  before_state: RegistryColumn | null;
  after_state: RegistryColumn | null;
  proposed_by_id: string;
  proposed_by_email: string;
  delta_version: number | null;
  apply_error: string | null;
  rejection_reason: string | null;
  created: string;
}

export interface ProposalDiffField {
  field: string;
  before: unknown;
  after: unknown;
}

export interface ProposalDetail extends Proposal {
  diff: ProposalDiffField[];
}

export interface DatasetVersion {
  version: number;
  timestamp: string;
  operation: string | null;
  proposal_id: string | null;
  approved_by_email: string | null;
}

export interface AuditLogEntry {
  id: string;
  proposal_id: string;
  action: string;
  actor_id: string;
  actor_email: string;
  created: string;
  dataset_key: string | null;
  column_name: string | null;
}

export interface DatasetImportError {
  row: number;
  detail: string;
}

export interface DatasetImportResponse {
  created: number;
  updated: number;
  skipped: number;
  errors: DatasetImportError[];
}

File: api/data_ingestion/schemas/schema_registry.py
from datetime import datetime
from typing import Any, Literal

from pydantic import UUID4, BaseModel, ConfigDict, Field

ProposalType = Literal["add", "edit", "delete"]
ProposalStatus = Literal["pending", "approved", "rejected", "apply_failed"]
CriticalFor = Literal["always", "create_only"]


class DatasetGroupCreate(BaseModel):
    key: str
    name: str
    description: str | None = None


class DatasetGroupResponse(DatasetGroupCreate):
    model_config = ConfigDict(from_attributes=True)

    id: str


class SchemaDatasetCreate(BaseModel):
    key: str
    group_id: str | None = None


class SchemaDatasetResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    key: str
    group_id: str | None = None


class SchemaDatasetSummary(SchemaDatasetResponse):
    column_count: int = 0
    pending_proposal_count: int = 0
    current_version: int | None = None


class MoveDatasetRequest(BaseModel):
    group_id: str | None = None


class DatasetLinkCreate(BaseModel):
    linked_dataset_key: str


class DatasetLinkResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    dataset_id: str
    linked_dataset_id: str


class RegistryColumn(BaseModel):
    """A column row as it exists on the Delta `schemas.<dataset>` table, extended
    with the DQ-config fields that used to live only in NocoDB."""

    model_config = ConfigDict(from_attributes=True)

    name: str
    data_type: str
    is_nullable: bool | None = None
    is_important: bool | None = None
    is_system_generated: bool | None = None
    description: str | None = None
    primary_key: bool | None = None
    partition_order: int | None = None
    license: str | None = None
    units: str | None = None
    hint: str | None = None
    is_mandatory: bool | None = None
    is_unique: bool | None = None
    domain_values: str | None = None
    range_min: float | None = None
    range_max: float | None = None
    dynamic_max: str | None = None
    precision_min: int | None = None
    # Only meaningful for master/geolocation datasets — see
    # giga-dagster's data_quality_checks/critical.py. Other dataset types
    # treat every mandatory column as automatically critical.
    critical_for: CriticalFor | None = None


class ProposalCreate(BaseModel):
    column_name: str
    proposal_type: ProposalType
    after_state: dict[str, Any] | None = None


class ProposalResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    dataset_id: str
    column_name: str
    proposal_type: ProposalType
    status: ProposalStatus
    before_state: dict[str, Any] | None = None
    after_state: dict[str, Any] | None = None
    proposed_by_id: UUID4
    proposed_by_email: str
    delta_version: int | None = None
    apply_error: str | None = None
    created: datetime


class ProposalDiffField(BaseModel):
    field: str
    before: Any = None
    after: Any = None


class ProposalDetail(ProposalResponse):
    diff: list[ProposalDiffField] = Field(default_factory=list)


class AuditLogResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    proposal_id: str
    action: str
    actor_id: UUID4
    actor_email: str
    created: datetime


class DatasetVersion(BaseModel):
    version: int
    timestamp: datetime
    operation: str | None = None
    proposal_id: str | None = None
    approved_by_email: str | None = None
